src/data/apigateway/mod.rs

```diff
@@ -0,0 +1,170 @@
+    DELETE,
+}
+
+impl fmt::Display for HttpMethod {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", match *self {
+            HttpMethod::HEAD => "HEAD",
+            HttpMethod::GET => "GET",
+            HttpMethod::POST => "POST",
+            HttpMethod::PUT => "PUT",
+            HttpMethod::OPTIONS => "OPTIONS",
+            HttpMethod::DELETE => "DELETE"
+        })
+    }
+}
+
+#[derive(Serialize,Deserialize)]
+pub struct HttpEvent {
+    pub body: Option<String>,
+    pub headers: Option<HashMap<String, String>>,
+    #[serde(rename="httpMethod")]
+    pub http_method: HttpMethod,
+    #[serde(rename="isBase64Encoded")]
+    pub is_base64_encoded: bool,
+    pub path: String,
+    #[serde(rename="pathParameters")]
+    pub path_parameters: Option<HashMap<String, String>>,
+    #[serde(rename="queryStringParameters")]
+    pub query_string_parameters: Option<HashMap<String, String>>,
+    pub resource: String,
+    #[serde(rename="requestContext")]
+    pub request_context: HttpEventRequestContext,
+    #[serde(rename="stageVariables")]
+    pub stage_variables: Option<HashMap<String, String>>,
+}
+
+impl HttpEvent {
+
+    pub fn get_header(&self, key: &str) -> Option<&str> {
+        match self.headers {
+            Some(ref h) => h.get(key).map(|s| s.as_str()),
+            None => None,
+        }
+    }
+}
+
+impl fmt::Display for HttpEvent {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{protocol} {method} {path}{querystring} (body? {body})",
+            protocol = match self.request_context.protocol {
+                Some(ref s) => s.as_str(),
+                None => "(unknown)"
+            },
+            method = self.http_method,
+            path = self.path,
+            querystring = qs::to_string(&self.query_string_parameters).unwrap_or(String::new()),
+            body = match self.body {
+                Some(_) => true,
+                None => false
+            }
+        )
+    }
+}
+
+#[derive(Serialize,Deserialize)]
+pub struct HttpEventRequestContext {
+    #[serde(rename="accountId")]
+    pub account_id: String,
+    #[serde(rename="apiId")]
+    pub api_id: String,
+    #[serde(rename="httpMethod")]
+    pub http_method: HttpMethod,
+    pub identity: HashMap<String, Option<String>>,
+    pub path: String,
+    pub protocol: Option<String>,
+    #[serde(rename="requestId")]
+    pub request_id: String,
+    #[serde(rename="requestTime")]
+    pub request_time: Option<String>,
+    #[serde(rename="requestTimeEpoch")]
+    pub request_time_epoch: Option<u64>,
+    #[serde(rename="resourceId")]
+    pub resource_id: String,
+    #[serde(rename="resourcePath")]
+    pub resource_path: String,
+    pub stage: String,
+}
+
+impl HttpEventRequestContext {
+
+    pub fn time(&self) -> Option<DateTime<Utc>> {
+        // Utc::datetime_from_str(&self.request_time, "").ok()e
+        None
+    }
+}
+
+serde_aux_enum_number_declare!(HttpStatus {
```

Comment:
I'd maybe suggest not going too overboard with type mappings here. I see crowbar being something as a low level mapping to raw primitive types, in this case a u16 would do. The reasoning behind this line of thinking is that you could then more easily adapt to and not duplicate efforts to create a typed interface like [http](https://github.com/hyperium/http/blob/master/src/status.rs#L305-L497) on top of crowbar. As is you couldn't build a mapping to the http crate on top of this because if the incomplete mapping of http statuses. I'd much prefer the crowbar retain the most primitive types and enable layering of extra layers on top in another crate.

Reply:
So should we rely on another crate for this? I do actually like having strongly typed things.

Reply:
I'd actually not recommend adding any more crate dependencies on crowbar. I don't see crowbar as a framework but as an faithful representation of lambda. I'm thinking now it its really api gateway, the service, that defines what status codes you can dispatch lambdas on so that may actually be a more restrictive set. I redact my previous comment!

Reply:
I agree that a u16 would do for the HTTP status.

I experimented a lot with the `http` crate for another project involving Lambda@Edge. It helped with initial development but I eventually ripped it out because it was doing a lot of validation that CloudFront already does for you.